### PR TITLE
Preserve indentation when writing JSON files

### PR DESCRIPTION
**Pull request by maximal**
Refs #11341, doesn’t resolve specifically it though, but has something to move on.

UPD: I’ll fix PHPStan warnings if the PR is OK in general.

### CODE REVIEW

**Comment by Seldaek** [comment]
Yeah that seems reasonable to me

File: src/Composer/Json/JsonFile.php [hunk 1]
```diff
@@ -262,15 +268,28 @@ public static function validateJsonSchema(string $source, $data, int $schema, ?s
...
         if (false === $json) {
             self::throwEncodeError(json_last_error());
         }
 
+        if (($options & JSON_PRETTY_PRINT) > 0 && $indent !== self::INDENT_DEFAULT ) {
+            // Pretty printing and not using default indentation
+            return Preg::replaceCallback(
+                '#^ {4,}#m',
+                static function ($match) use ($indent): string {
+                    return str_repeat($indent, (int)(strlen($match[0] ?? '') / 4));
```
Comment (stof) [inline src/Composer/Json/JsonFile.php:287]: if the number of spaces is not a multiple of 4, we should probably keep the remaining spaces (as those would not be about indentation but about alignment)

**Review by stof — COMMENTED** [review]
it would be great to add tests covering that behavior

File: src/Composer/Json/JsonFile.php (hunk 1 above)
Comment (maximal) [reply src/Composer/Json/JsonFile.php:287]: Mathematically speaking, number of tabs must always be equal of indent level, and then any number of spaces for alignment (aka [SmartTabs](https://www.emacswiki.org/emacs/SmartTabs) approach), but I think here we can simplify the process as we are dealing with `json_encode()` result which can have either 4 or 0 spaces for indentation and have no alignment at all.

**Comment by Seldaek** [comment]
Thanks

**Pull request merged**